File: python/95_Unique_Binary_Search_Trees_II.py
#!/usr/bin/env python3
import copy
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        result = []


    def helper(self, root, curRoot, curIndex, maxIndex, resultList):
        if curIndex == maxIndex:
            resultList.append(root)
            return
        leftRoot = TreeNode()
        rightRoot = TreeNode()

        leftCur = self.myCopy(root, leftRoot, curRoot)
        rightCur = self.myCopy(root, rightRoot, curRoot)

        


    def myCopy(self, root, newRoot, target):
        newRoot.val = root.val
        result1 = result2 = None
        if root.left is not None:
            newRoot.left = TreeNode()
            result1 = self.myCopy(root.left, newRoot.left, target)
        if root.right is not None:
            newRoot.right = TreeNode()
            result2 = self.myCopy(root.right, newRoot.right, target)

        if root == target:
            return newRoot
        elif result1 is not None:
            return result1
        elif result2 is not None:
            return result2
        else:
            return None














s = Solution()

a = TreeNode('a')
b = TreeNode('b')
c = TreeNode('c')
a.right = b
b.left = c

File: java/src/brand_new/_95_Unique_Binary_Search_Trees_II.java
package brand_new;

import java.util.ArrayList;
import java.util.List;
/*
 * 还可以用动态规划优化
 * 下面这个写法，部分树节点是共用的
 */
public class _95_Unique_Binary_Search_Trees_II {

	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode(int x) {
			val = x;
		}
	}

	public List<TreeNode> generateTrees(int n) {
		if(n == 0)
			return new ArrayList<TreeNode>();
		return generateTrees(1, n);
	}
	
	public List<TreeNode> generateTrees(int min, int max){
		List<TreeNode> result = new ArrayList<TreeNode>();
		if(min > max){
			result.add(null);
			return result;
		}
		
		if(min == max){
			result.add(new TreeNode(min));
			return result;
		}
		
		for(int i = min; i <= max; i++){
			List<TreeNode> left = generateTrees(min, i - 1);
			List<TreeNode> right = generateTrees(i+1, max);
			
			for(int j = 0; j < left.size(); j++){
				for(int k = 0; k < right.size(); k++){
					TreeNode node = new TreeNode(i);
					node.left = left.get(j);// deep copy !!!
					node.right = right.get(k);
					result.add(node);
				}
			}
		}
		return result;
	}
	

	public static void main(String[] args) {
		List<TreeNode> result = new _95_Unique_Binary_Search_Trees_II().generateTrees(3);
		System.out.println(result.size());
	}

}
